File: JSClassifier/featureExtractor.js
const htmlParser =  require('htmlparser2');

/*
    Uso:
        let extractor = new FeatureExtractor('www.niqq.com.br', <html_da_pagina>);
        let feaures = extractor.getFeatures();
*/
class FeatureExtractor{
    constructor(url, html){
        this.features = {};
        this.html = html;
        this.url = url;
        this.parser = this.getParser();
        
        this.urlRegex = ['regist', 'cadast', 'client', 'sign.?up', ];
        this.inputTypes = ["password", "email"];
        this.inputRegex = ["n[ao]me", "e?.?mail", "password", "senha",
                             "phone", "number", "user", "address", "cep"];
        this.attributesSearched = ["name", "placeholder", "id", "class"];
        this.action_url_regex = ["create", "new","nov[oa]", "cadast", "edit", "salvar", "save"];

        this.prepare();
    }

    prepare(){
        this.initializeFeatureObject("url", this.urlRegex);
        this.initializeFeatureObject("inputs_type_", this.inputTypes);
        this.initializeFeatureObject("inputs_reg_", this.inputRegex);
        this.initializeFeatureObject("href_auth_regex_", this.action_url_regex);
    }
    
    initializeFeatureObject(prefix, regexList){
        regexList.forEach(function(regex){
            let key = prefix + regex;
            this.features[key] = 0;
        }, this);
    }

    getParser(){
        let self = this;
        return new htmlParser.Parser({
            onopentag: function(name, attrs){
                if(name === "input")
                    self.parseInput(attrs);
                if(name === "a")
                    self.processLink(attrs);
            }
        }, { decodeEntities: true });
    }
    
    // 'g' -> match as much as possible
    // 'i' -> ignore case
    matchCount(pattern, str){
        if(!str)
            return 0;
        let matches = str.match(new RegExp(pattern, 'gi')) || [];
        return matches.length;
    }
    
    extractUrlFeaturesFrom(url){
        this.urlRegex.forEach(function(feature) {
            this.features['url' + feature] = this.matchCount(feature, url);
        }, this);
    }
    
    parseInputType(type){
        let key = "inputs_type_" + type;
    
        if(this.inputTypes.indexOf(type) > -1){
            this.features[key] += 1;
        }
    }
    
    processInputAttr(infoText){
        this.inputRegex.forEach(function(regex){
            let key = "inputs_reg_" + regex;
            this.features[key] += this.matchCount(regex, infoText);
        }, this);
    }
    
    parseInput(attrs){
        let infoText = "";
        if(attrs.type)
            this.parseInputType(attrs.type);
        
        Object.keys(attrs).forEach(function(key){
            if(this.attributesSearched.indexOf(key) > -1){
                infoText += attrs[key];
            }
        }, this);
    
        if(infoText !== "")
            this.processInputAttr(infoText);
    }
    
    /*
        Retorna todas as features do site passado no construtor.
    */
    getFeatures(){
        let self = this;
        return new Promise((resolve, reject) => {
            self.parser.write(self.html);
            self.extractUrlFeaturesFrom(self.url);
            let binFeatures = {};
            Object.keys(self.features).forEach(function(key){
                console.log(key);
                binFeatures[key] = self.features[key] > 0 ? 1 : 0;
            }, this);
            resolve(binFeatures);
        });
    }
}

module.exports.featureExtractor = FeatureExtractor;

File: TensorFlowCNN/Prototype2/feature_extractor.py
from HTMLParser import HTMLParser
from selenium import webdriver
import re

def urlFeatures(url):
    result = dict()
    url_regex = ["regist", "cadast", "client", "sign.?up"]
    for reg in url_regex:
        result['url' + reg] = len(re.findall(reg, url, re.IGNORECASE))
    return result


class HTMLFeatExtract(HTMLParser):
    inputs_type = ["password", "email"]
    inputs_regex = ["first", "last", "n[ao]me", "e?.?mail", "password", "senha", "date", "phone", "number", "user", "address",
                        "cep", "cpf","uf","complemento","numero","endereco"]
    href_regex = ["term", "privac"]
    href_auth_regex = ["facebook", "google"]
    attributes_searched = ["name", "placeholder", "id", "class"]
    action_url_regex = ["create", "new","nov[oa]", "cadast", "edit", "salvar", "save"]
    
    def __init__(self):
        HTMLParser.__init__(self)

        self.feats = dict()
        self.number_of_inputs = 0
        for kind in self.inputs_type:
            self.feats["inputs_type_" + kind] = 0
        for reg in self.inputs_regex:
            self.feats["inputs_reg_" + reg] = 0
        for href in self.href_regex:
            self.feats['href_reg_' + href] = 0
        for href_auth in self.href_auth_regex:
            self.feats['href_auth_regex_' + href_auth] = 0
        for pattern in self.action_url_regex:
            self.feats['form_action_' + pattern] = 0

    def getFeatures(self):
        return self.feats

    def handle_starttag(self, tag, attrs):
        if tag == "form":
            self.process_form(attrs)
        if tag == "input":
            self.number_of_inputs += 1
            self.process_input(attrs)
        if tag == "a":
            self.process_link(attrs)

    def handle_endtag(self, tag):
        pass
    
    def handle_data(self, data):
        pass

    def process_form(self, attrs):
        for att in attrs:
            if(att[0] == "action"):
                for reg in self.action_url_regex:
                    self.feats['form_action_' + reg] += len(re.findall(reg, att[1], re.IGNORECASE))

    def process_link(self, attrs):
        for att in attrs:
            if att[0] == "href":
                for reg in self.href_regex:
                    self.feats['href_reg_' + reg] += len(re.findall(reg, att[1], re.IGNORECASE))
                if (re.findall("auth", att[1], re.IGNORECASE)):
                    for reg in self.href_auth_regex:
                        self.feats['href_auth_regex_' + reg] += len(re.findall(reg, att[1], re.IGNORECASE))

    def process_input(self, attrs):
        info_text = ""
        for att in attrs:
            if att[0] == "type":
                self.process_input_type(att[1])
            if att[0] in self.attributes_searched:
                info_text += att[1]
        self.process_input_attr(info_text)

    def process_input_type(self, intype):
        for kind in self.inputs_type:
            if intype == kind:
                self.feats["inputs_type_" + kind] += 1

    def process_input_attr(self, inname):
        for reg in self.inputs_regex:
            self.feats["inputs_reg_" + reg] += len(re.findall(reg, inname, re.IGNORECASE)) 


def getAllFeatures(url, data):
    extract = HTMLFeatExtract()
    
    feat = urlFeatures(url)
    extract.feed(data)
    feat.update(extract.getFeatures())

    return feat, extract.number_of_inputs